library/Fission/Web/Heroku/Deprovision.hs

```diff
@@ -50,73 +55,39 @@ deleteAssociatedWith ::
 
-  let cids = getInner userCidCid <$> userCids
-  remaining <- getRemainingCIDs cids
+  deleteAssociatedRecords userId uuid' userCids
 
-  let remainingCIDs = getInner userCidCid <$> remaining
-  return (cids \\ remainingCIDs)
+  let deletedUserCids = getInner userCidCid <$> userCids
+  remainingUserCids <- UserCIDQuery.getByCids deletedUserCids
 
--- | Find all CIDs that remain from a list
-getRemainingCIDs :: MonadDB m => [CID] -> Transaction m [Entity UserCid]
-getRemainingCIDs cids =
-  select <| from \userCid -> do
-    where_ (userCid ^. UserCidCid `in_` valList cids)
-    return userCid
+  let remainingCIDs = getInner userCidCid <$> remainingUserCids
+  return (deletedUserCids \\ remainingCIDs)
 
 -- | All records associated with the UUID, across the user, user CID, and Heroku add-on tables
-associatedRecords :: UUID -> [Entity UserCid] -> SqlQuery ()
-associatedRecords uuid' userCids = do
-  let userCidIds = entityKey <$> userCids
-  from \userCid ->
-    where_ (userCid ^. UserCidId `in_` valList userCidIds)
-
-  let userIds' = getInner userCidUserFk <$> userCids
-  from \user ->
-    where_ (user ^. UserId `in_` valList userIds')
-
-  from \herokuAddOn ->
-    where_ (herokuAddOn ^. HerokuAddOnUuid ==. val uuid')
-
--- | CIDs associated with a user
-cidsForUserId :: MonadDB m => UserId -> Transaction m [Entity UserCid]
-cidsForUserId userId =
-  select <| from \userCid -> do
-    where_ (userCid ^. UserCidUserFk ==. val userId)
-    return userCid
-
--- | Users associated with those Heroku add-ons
+deleteAssociatedRecords :: MonadDB m => UserId -> UUID -> [Entity UserCid] -> Transaction m ()
+deleteAssociatedRecords userId uuid userCids = do
+  UserCIDMutation.destroyAll      (entityKey <$> userCids)
+  UserMutation.destroy        userId
+  UserMutation.destroyHerokuAddon uuid -- TODO would be nice if this was cascading....
+
+-- | Get the User associated with those Heroku add-ons, throw 410 if not found.
 userIdForHerokuAddOn ::
   ( MonadDB     m
   , MonadLogger m
   , MonadThrow  m
   )
   => HerokuAddOnId
   -> Transaction m UserId
-userIdForHerokuAddOn addOnId = ensureOneId err410 =<< query
-  where
-    query = select <| from \user -> do
-      where_ <| user ^. UserHerokuAddOnId ==. val (Just addOnId)
-            &&. user ^. UserActive        ==. val True
-      limit 1
-      return user
-
--- | Heroku add-on with a specific UUID
+userIdForHerokuAddOn addOnId = ensureOneId err410 =<< (UserQuery.getHerkouAddonByUserId addOnId)
+
+-- | Get a Heroku add-on with a specific UUID, throw 410 if not found.
 herokuAddOnByUUID ::
   ( MonadDB     m
   , MonadLogger m
   , MonadThrow  m
   )
   => UUID
   -> Transaction m HerokuAddOnId
-herokuAddOnByUUID uuid' = ensureOneId err410 =<< query
-  where
-    query = select <| from \herokuAddOn -> do
-      where_ (herokuAddOn ^. HerokuAddOnUuid ==. val uuid')
-      limit 1
-      return herokuAddOn
+herokuAddOnByUUID uuid' = ensureOneId err410 =<< (UserQuery.getHerkouAddonByUUID uuid')
```

Comment:
💅 No need for the parens:

```haskell
herokuAddOnByUUID uuid' = ensureOneId err410 =<< UserQuery.getHerkouAddonByUUID uuid'
```

Or if you want pipes (though I personally find this less clear)

```haskell
herokuAddOnByUUID uuid' =
  uuid'
    |> UserQuery.getHerkouAddonByUUID
    |> bind (ensureOneId err410)
```